neuralprophet/time_net.py

```diff
@@ -767,6 +830,165 @@ def compute_components(self, inputs):
+        return loss, reg_loss
+
+    def training_step(self, batch, batch_idx):
+        inputs, targets, _ = batch
+        # Run forward calculation
+        predicted = self.forward(inputs)
+        # Calculate loss
+        loss, reg_loss = self.loss_func(inputs, predicted, targets)
+        # Metrics
+        if not self.minimal:
+            predicted_denorm = self.denormalize(predicted[:, :, 0])
+            target_denorm = self.denormalize(targets.squeeze(dim=2))
+            self.log_dict(self.metrics_train(predicted_denorm, target_denorm), **self.log_args)
+            self.log("Loss", loss, **self.log_args)
+            self.log("RegLoss", reg_loss, **self.log_args)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        inputs, targets, _ = batch
+        # Run forward calculation
+        predicted = self.forward(inputs)
+        # Calculate loss
+        loss, reg_loss = self.loss_func(inputs, predicted, targets)
+        # Metrics
+        if not self.minimal:
+            predicted_denorm = self.denormalize(predicted[:, :, 0])
+            target_denorm = self.denormalize(targets.squeeze(dim=2))
+            self.log_dict(self.metrics_val(predicted_denorm, target_denorm), **self.log_args)
+            self.log("Loss_val", loss, **self.log_args)
+            self.log("RegLoss_val", reg_loss, **self.log_args)
+
+    def test_step(self, batch, batch_idx):
+        inputs, targets, _ = batch
+        # Run forward calculation
+        predicted = self.forward(inputs)
+        # Calculate loss
+        loss, reg_loss = self.loss_func(inputs, predicted, targets)
+        # Metrics
+        self.log("Loss_test", loss, **self.log_args)
+        self.log("RegLoss_test", reg_loss, **self.log_args)
+
+    def predict_step(self, batch, batch_idx, dataloader_idx=0):
+        inputs, _, _ = batch
+        # Add predict_mode flag to dataset
+        inputs["predict_mode"] = True
+        # Run forward calculation
+        prediction = self.forward(inputs)
+        # Calculate components (if requested)
+        if self.compute_components_flag:
+            components = self.compute_components(inputs)
+        else:
+            components = None
+        return prediction, components
+
+    def configure_optimizers(self):
+        # Optimizer
+        optimizer = self.optimizer(self.parameters(), lr=self.learning_rate, **self.config_train.optimizer_args)
+
+        # Scheduler
+        steps_per_epoch = math.ceil(self.trainer.estimated_stepping_batches / self.trainer.max_epochs)
+        lr_scheduler = {
+            "scheduler": self.scheduler(
+                optimizer,
+                max_lr=self.learning_rate,
+                epochs=self.trainer.max_epochs,
+                steps_per_epoch=steps_per_epoch,
+                **self.config_train.scheduler_args,
+            ),
+            "name": "learning_rate",
+            "interval": "step",
+        }
+
+        return [optimizer], [lr_scheduler]
```

Comment:
Info: The optimizer and scheduler are only initialized here, since `self.learning_rate` is accessed and tuned by the learning rate scheduler. Further, args such as the number of steps per epoch are only available at runtime in the trainer object and not before the model is trained.